### PR TITLE
Take focus from game during calls on iOS

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes ppy/osu#4565
- Split from https://github.com/ppy/osu-framework/pull/6449

Again, this is not normal behaviour of any iOS app, as incoming calls have recently been acting more like a glorified notification than a full-screen overlay. However, as a rhythm game, exclusive access to the audio is required for the game to function appropriately, therefore we have to observe for incoming calls and cut focus from the game in order to be aware that it should not remain operating (i.e. send to pause screen).

Note that this change is made primarily because users would otherwise lose their scores when they get a call, as the beatmap freezes and osu! triggers the "you have an issue with your audio device, score will not be submitted" notification.

The framework may not be exclusively for rhythm games, but I think it's simple to have this anyway until we get a report about it from other framework consumers.

### CODE REVIEW

**Comment by peppy** [comment]
To clarify, this is avoiding the game becoming unusable when a phone call arrives? Because your description sounds like this is an optional change, and not having it would just mean that the phone call notification shows on top of the game.

Maybe you could show a video of what happens with/without this to clarify.

**Comment by frenzibyte** [comment]
I briefly mentioned it in https://github.com/ppy/osu-framework/pull/6449#issue-2727329289 but forgot to mention here. This change is mandatory for osu!, otherwise when someone receives a call the beatmap freezes and osu! thinks the audio system is at fault and excludes the score from submission.

**Comment by peppy** [comment]
@frenzibyte I found the handling of sdl events locally in `IOSWindow` pretty ugly, so I've refactored a bit to avoid this. It's still not pretty, but it's the best I can come up with. cc @Susko3 for a quick check.

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
Requesting a second set of eyes on this (low priority) @ppy/team-client.

File: osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs [hunk 1]
```diff
@@ -122,22 +122,10 @@ private void initialiseWindowingAfterCreation()
...
-            get => focused;
-            protected set
-            {
-                if (value == focused)
-                    return;
-
-                isActive.Value = focused = value;
-            }
-        }
+        public bool Focused { get; private set; }
```
Comment (Susko3) [inline osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs:128]:
```suggestion
        public bool Focused { get; protected set; }
```

To fix Android build.

File: osu.Framework.iOS/IOSWindow.cs [hunk 2]
```diff
@@ -47,8 +49,31 @@ public override void Create()
...
+            UpdateActiveState();
         }
 
+        private void onCallEnded()
+        {
+            inCall = false;
+            UpdateActiveState();
+        }
+
+        protected override bool ShouldBeActive => base.ShouldBeActive && !inCall;
```
Comment (Susko3) [inline osu.Framework.iOS/IOSWindow.cs:75]: While it fixes the issue, I think that forcing the game to be inactive when a call is in progress isn't fully correct.

Image this scenario:
1. Play osu!
2. Get a call
   - game is paused as expected
3. Answer call (so that osu! goes to the background)
4. Go back to osu!
   - game is focused, but `IsActive == false`, leading to weird behaviour (likely: input doesn't work, the game cannot be unpaused)
5. End call
6. Go back to osu!
   - everything is as expected

Instead of the game checking if a call is in progress, I think it should check if a call is currently ringing. Looking at the [apple docs](https://developer.apple.com/documentation/callkit/cxcall), I would consider a call to be ringing iff:
- `!isOutgoing &&`
- `!hasConnected &&`
- `!hasEnded &&`
- `!isOnHold`.

I think it's fine to force the game to be inactive when a call is ringing, as the user is focused on answering or declining the call.

**Review by Susko3 — CHANGES_REQUESTED** [review]

**Comment by Susko3** [comment]
Everything else seems fine to me.

File: osu.Framework.iOS/IOSWindow.cs (hunk 2 above)
Comment (peppy) [reply osu.Framework.iOS/IOSWindow.cs:75]: cc @frenzibyte, looks better than what you have.

File: osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs (hunk 1 above)
Comment (peppy) [reply osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs:128]: I'd rather fix this by updating android code if we can.

File: osu.Framework.iOS/IOSWindow.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Framework.iOS/IOSWindow.cs:75]: This would be more optimal, but the audio system always remains frozen while a call is active, from the point of ringing until the call is ended. Therefore it doesn't help to make the game focused while a call is active (the user would be faced with a frozen beatmap and they will receive the "audio is not working" message etc.).

If anything, we should prevent the user from trying to unpause the game during a call, but that's pretty much follow-up effort for this PR.

Comment (Susko3) [reply osu.Framework.iOS/IOSWindow.cs:75]: > the audio system always remains frozen while a call is active

😳 Is this how other games behave? If not,  maybe something is wrong with BASS.

Comment (frenzibyte) [reply osu.Framework.iOS/IOSWindow.cs:75]: Initial testing with Arcaea showed much proper handling. When an incoming call is triggered, the level still continues progressing, the audio stops for a short time but returns back, and after accepting a call the audio remains working.

I'll report this to BASS.

Comment (peppy) [reply osu.Framework.iOS/IOSWindow.cs:75]: Honestly let's just go with what we have for now. The above scenario is an edge case of an edge case. We can iterated.

File: osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs (hunk 1 above)
Comment (peppy) [reply osu.Framework/Platform/SDL3/SDL3Window_Windowing.cs:128]: @Susko3 can you test the android change i committed?
